CBinding/Project/CProject.cs

```diff
@@ -437,50 +524,81 @@ protected override SolutionItemConfiguration OnCreateConfiguration (string name,
 				packages.Project = this;
 			}
 		}
-		
+
+		/// <summary>
+		/// Invoked when a file is added to project. Detects UTF-8 BOM.
+		/// </summary>
+		/// <param name="args">Arguments.</param>
 		protected override void OnFileAddedToProject (ProjectFileEventArgs args)
 		{
 			base.OnFileAddedToProject (args);
-			
+			foreach(var f in Files){
+				using (var s = new FileStream (f.Name, FileMode.Open, FileAccess.Read, FileShare.ReadWrite)) {
+					byte[] BOM = new byte[3];
+					s.Read (BOM, 0, 3);
+					if (!BOMPresentInFile.ContainsKey (f.Name)) {
+						BOMPresentInFile.Add (f.Name, false);
+					}
+					if(BOM[0] == 0xEF && BOM[1] == 0xBB && BOM[2] == 0xBF){
+						BOMPresentInFile[f.Name] = true;
+					} else {
+						BOMPresentInFile[f.Name] = false;
+					}
+				}
+			}
 			foreach (ProjectFileEventInfo e in args) {
 				if (!Loading && !IsCompileable (e.ProjectFile.Name) &&
-				    e.ProjectFile.BuildAction == BuildAction.Compile) {
+					e.ProjectFile.BuildAction == BuildAction.Compile) {
 					e.ProjectFile.BuildAction = BuildAction.None;
 				}
-				
+
 				if (e.ProjectFile.BuildAction == BuildAction.Compile)
-					TagDatabaseManager.Instance.UpdateFileTags (this, e.ProjectFile.Name);
+					ThreadPool.QueueUserWorkItem (o => {
```

Comment:
This looks really hacky, it's going to have thread safety and race issues. Would be much better to use the type system service's parser queue.